Task#4 Container

KroosT opened the pull request:
Type-safe container which can store values of types that implements AnswersTo interface.
Added method allThatAnswerTo() returns new instance of container but only with values that answers certain request which can be any type.

KroosT committed:
Task #4 Container moved from another branch.

KroosT committed:
Added Util Class Validator.

Contains isArgNotNull() method that checks if passed argument is null and throw exception.

KroosT committed:
.gitignore changed

vmatelsky commented on Tasks/src/ContainerTask3/Container.java:
```diff
@@ -0,0 +1,56 @@
+
+    Container(final Container<T, E> container) {
+        Validator.isArgNotNull(container, "container");
+        mList = new ArrayList<>(container.getList());
```
double copying

vmatelsky commented on Tasks/src/ContainerTask3/Container.java:
```diff
@@ -0,0 +1,56 @@
+        mList.add(object);
+    }
+
+    T remove(final int index) {
```
naming

vmatelsky commented on Tasks/src/ContainerTask3/ContainerTest.java:
```diff
@@ -0,0 +1,31 @@
+
+public class ContainerTest {
+    @Test
+    public void allThatAnswersTo() throws Exception {
```
split into two methods

KroosT committed:
Classes modified.

Got rid of double copying.
Renamed test method to more descriptive.
Split method into two.

vmatelsky commented on Tasks/src/ContainerTask3/ContainerTest.java:
```diff
@@ -8,24 +8,37 @@
+        final Container<Realtek, String> realtekContainerAnswersTo = realtekContainer
+                .allThatAnswersTo("AUDIO");
+
+        Assert.assertEquals(realtekContainerAnswersTo.getList().size(), 3);
```
can import static methods of `Assert`:
```
assertEquals(realtekContainerAnswersTo.getList().size(), 3);
```

use AssertJ for clearness

KroosT committed:
Renamed remove methods.

KroosT committed:
Improved tests by using AssertJ.

KroosT committed:
Changed isEqualsTo() on containsExactlyElementsOf()
